crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3427,9 +3472,16 @@ impl<'db> TypeInferenceBuilder<'db> {
                             ),
                         );
                     }
-                }
 
-                Type::Never
```

Comment:
Is it important that we return `Never` here? I wasn't able to construct a Python example that would get access to this type...?

Reply:
It was changed to use `Never` in https://github.com/astral-sh/ruff/commit/b1ce8a39495918d1b7421317b8b8e63d0653d3df, which was prompted by https://github.com/astral-sh/ruff/pull/13981#issuecomment-2445472433

Reply:
You can't get access to this type in Python, the only reason we need a type for it is the "type pulling" test, which is a proxy for "what should we show if you hover over this expression in an IDE". `Never` is my best answer for that (I think it's a better answer than `None`), but I don't think it matters much.

Reply:
> You can't get access to this type in Python, the only reason we need a type for it is the "type pulling" test, which is a proxy for "what should we show if you hover over this expression in an IDE". `Never` is my best answer for that (I think it's a better answer than `None`), but I don't think it matters much.

Sorry, I should have phrased my question better: *given* that we "don't think it matters much", is it okay if we change the type of the `obj.attr` expression in an assignment like `obj.attr = value` from `Never` to the type that we would get for `obj.attr` in a `Load` context? I'm inclined to say that this is much more helpful for a user in the IDE, which is why I thought there must have been a reason why `Never` was chosen instead.

Reply:
To get this merged, I'm going to assume that it's okay to do this (let me know if not!).

Reply:
In a case like `obj.attr = value`, I think it's fine if hover on `obj.attr` shows the type of `value`. If it shows the type of `obj.attr` prior to the assignment, I think that's not good. (I haven't had a chance to fully review this PR yet so I don't know yet which it is.)

Reply:
> In a case like `obj.attr = value`, I think it's fine if hover on `obj.attr` shows the type of `value`. If it shows the type of `obj.attr` prior to the assignment, I think that's not good.

Okay. In that case, this will need another iteration (it's the latter).

Reply:
I just realized that until we add narrowing on assignments to attributes, there's no difference in the `obj.attr = value` case. But in the `obj = value`, or particularly the `obj: annotation = value` case, it could be a totally different type, and I don't think it's sensible for hover on that assignment target to show its type prior to the assignment.

Reply:
But this change only applies to attribute expressions, so for now at least I think it's fine as-is.